File: lib/src/apps/quran/core/entities/audio_entities.dart
import 'package:equatable/equatable.dart';

/// كيان القارئ
class ReciterEntity extends Equatable {
  final int id;
  final String? name;
  final bool isComplete;
  final int? approvalStatus;
  final int? countryId;
  final String? profile;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const ReciterEntity({
    required this.id,
    this.name,
    this.isComplete = false,
    this.approvalStatus,
    this.countryId,
    this.profile,
    this.createdAt,
    this.updatedAt,
  });

  @override
  List<Object?> get props => [id, name, isComplete, approvalStatus, countryId, profile];
}

/// كيان التلاوة/الرواية
class RecitationEntity extends Equatable {
  final int id;
  final String name;
  final bool isActive;
  final DateTime? createdAt;

  const RecitationEntity({
    required this.id,
    required this.name,
    this.isActive = false,
    this.createdAt,
  });

  @override
  List<Object?> get props => [id, name, isActive];
}

/// كيان نوع القارئ (ربط القارئ بالتلاوة)
class ReciterTypeEntity extends Equatable {
  final int id;
  final int? reciterId;
  final int? recitationId;
  final int? approvalStatus;

  const ReciterTypeEntity({
    required this.id,
    this.reciterId,
    this.recitationId,
    this.approvalStatus,
  });

  @override
  List<Object?> get props => [id, reciterId, recitationId, approvalStatus];
}

/// كيان ملف صوتي للسورة
class ReciterAudioEntity extends Equatable {
  final int id;
  final int? reciterId;
  final int? surahId;
  final int? reciterTypeId;
  final String audio;
  final int? approvalStatus;
  final bool isComplete;

  const ReciterAudioEntity({
    required this.id,
    this.reciterId,
    this.surahId,
    this.reciterTypeId,
    required this.audio,
    this.approvalStatus,
    this.isComplete = false,
  });

  @override
  List<Object?> get props => [id, reciterId, surahId, reciterTypeId, audio, isComplete];
}

/// كيان صوت الآية (للتشغيل المتواصل)
class AyahAudioEntity extends Equatable {
  final int id;
  final int? reciterAudioId;
  final int? reciterId;
  final int? surahId;
  final int? ayahId;
  final int? ayahOrder;
  final int? reciterTypeId;
  final String audio;
  final int? startTime; // بالمللي ثانية
  final int? endTime; // بالمللي ثانية

  const AyahAudioEntity({
    required this.id,
    this.reciterAudioId,
    this.reciterId,
    this.surahId,
    this.ayahId,
    this.ayahOrder,
    this.reciterTypeId,
    required this.audio,
    this.startTime,
    this.endTime,
  });

  Duration? get startDuration =>
      startTime != null ? Duration(milliseconds: startTime!) : null;
  Duration? get endDuration =>
      endTime != null ? Duration(milliseconds: endTime!) : null;

  @override
  List<Object?> get props => [
        id,
        reciterAudioId,
        reciterId,
        surahId,
        ayahId,
        ayahOrder,
        reciterTypeId,
        audio,
        startTime,
        endTime,
      ];
}

File: huda_allah_system/apps/backend/src/database/entities/audio.entities.ts
import { Entity, PrimaryGeneratedColumn, Column, ManyToOne, OneToMany, JoinColumn } from 'typeorm';

@Entity('quran_reciter')
export class Reciter {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'deleted_at', nullable: true, type: 'datetime' })
  deletedAt: Date;

  @Column({ name: 'created_at', nullable: true, type: 'datetime' })
  createdAt: Date;

  @Column({ name: 'updated_at', nullable: true, type: 'datetime' })
  updatedAt: Date;

  @Column({ nullable: true, type: 'text' })
  name: string;

  @Column({ name: 'is_complete', default: false })
  isComplete: boolean;

  @Column({ name: 'approval_status', nullable: true })
  approvalStatus: number;

  @Column({ name: 'country_id', nullable: true, type: 'bigint' })
  countryId: number;

  @Column({ length: 100, nullable: true })
  profile: string;

  // Relations
  @OneToMany(() => ReciterType, reciterType => reciterType.reciter)
  reciterTypes: ReciterType[];

  @OneToMany(() => ReciterAudio, reciterAudio => reciterAudio.reciter)
  reciterAudios: ReciterAudio[];
}

@Entity('quran_recitation')
export class Recitation {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'deleted_at', nullable: true, type: 'datetime' })
  deletedAt: Date;

  @Column({ name: 'created_at', nullable: true, type: 'datetime' })
  createdAt: Date;

  @Column({ name: 'updated_at', nullable: true, type: 'datetime' })
  updatedAt: Date;

  @Column({ length: 50 })
  name: string;

  @Column({ name: 'is_active', default: false })
  isActive: boolean;

  // Relations
  @OneToMany(() => ReciterType, reciterType => reciterType.recitation)
  reciterTypes: ReciterType[];
}

@Entity('quran_recitertype')
export class ReciterType {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'deleted_at', nullable: true, type: 'datetime' })
  deletedAt: Date;

  @Column({ name: 'created_at', nullable: true, type: 'datetime' })
  createdAt: Date;

  @Column({ name: 'updated_at', nullable: true, type: 'datetime' })
  updatedAt: Date;

  @Column({ name: 'reciter_id', nullable: true, type: 'bigint' })
  reciterId: number;

  @Column({ name: 'recitation_id', nullable: true, type: 'bigint' })
  recitationId: number;

  @Column({ name: 'approval_status', nullable: true })
  approvalStatus: number;

  // Relations
  @ManyToOne(() => Reciter, reciter => reciter.reciterTypes)
  @JoinColumn({ name: 'reciter_id' })
  reciter: Reciter;

  @ManyToOne(() => Recitation, recitation => recitation.reciterTypes)
  @JoinColumn({ name: 'recitation_id' })
  recitation: Recitation;

  @OneToMany(() => ReciterAudio, reciterAudio => reciterAudio.reciterType)
  reciterAudios: ReciterAudio[];
}

@Entity('quran_reciteraudio')
export class ReciterAudio {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'deleted_at', nullable: true, type: 'datetime' })
  deletedAt: Date;

  @Column({ name: 'created_at', nullable: true, type: 'datetime' })
  createdAt: Date;

  @Column({ name: 'updated_at', nullable: true, type: 'datetime' })
  updatedAt: Date;

  @Column({ name: 'reciter_id', nullable: true, type: 'bigint' })
  reciterId: number;

  @Column({ name: 'surah_id', nullable: true, type: 'bigint' })
  surahId: number;

  @Column({ name: 'reciter_type_id', nullable: true, type: 'bigint' })
  reciterTypeId: number;

  @Column({ length: 100 })
  audio: string;

  @Column({ name: 'approval_status', nullable: true })
  approvalStatus: number;

  @Column({ name: 'is_complete', default: false })
  isComplete: boolean;

  // Relations
  @ManyToOne(() => Reciter, reciter => reciter.reciterAudios)
  @JoinColumn({ name: 'reciter_id' })
  reciter: Reciter;

  @ManyToOne(() => ReciterType, reciterType => reciterType.reciterAudios)
  @JoinColumn({ name: 'reciter_type_id' })
  reciterType: ReciterType;

  @OneToMany(() => AyahAudio, ayahAudio => ayahAudio.reciterAudio)
  ayahAudios: AyahAudio[];
}

@Entity('quran_ayahaudio')
export class AyahAudio {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'deleted_at', nullable: true, type: 'datetime' })
  deletedAt: Date;

  @Column({ name: 'created_at', nullable: true, type: 'datetime' })
  createdAt: Date;

  @Column({ name: 'updated_at', nullable: true, type: 'datetime' })
  updatedAt: Date;

  @Column({ name: 'reciter_audio_id', nullable: true, type: 'bigint' })
  reciterAudioId: number;

  @Column({ name: 'reciter_id', nullable: true, type: 'bigint' })
  reciterId: number;

  @Column({ name: 'surah_id', nullable: true, type: 'bigint' })
  surahId: number;

  @Column({ name: 'ayah_id', nullable: true, type: 'bigint' })
  ayahId: number;

  @Column({ name: 'ayah_order', nullable: true, type: 'int', unsigned: true })
  ayahOrder: number;

  @Column({ name: 'reciter_type_id', nullable: true, type: 'bigint' })
  reciterTypeId: number;

  @Column({ length: 100 })
  audio: string;

  @Column({ name: 'start_time', nullable: true, type: 'bigint', unsigned: true })
  startTime: number; // milliseconds

  @Column({ name: 'end_time', nullable: true, type: 'bigint', unsigned: true })
  endTime: number; // milliseconds

  // Relations
  @ManyToOne(() => ReciterAudio, reciterAudio => reciterAudio.ayahAudios)
  @JoinColumn({ name: 'reciter_audio_id' })
  reciterAudio: ReciterAudio;
}
